File: Leetcode/142.Linked_List_CycleII/C++/142Linked_List_CycleII.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 18 22:55:25 2023

@author: User
"""
from typing import Optional
class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None
         
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        hashTable={}
        dummyhead=ListNode(-10001)
        dummyhead.next=head
        ptr=dummyhead.next
        while ptr!=None:
            if ptr in hashTable and hashTable[ptr]==1:
                return ptr
            hashTable[ptr]=1
            ptr=ptr.next
        return None

sol=Solution()
head=ListNode(3)
head.next=ListNode(2)
temp=head.next
head.next.next=ListNode(0)
head.next.next.next=ListNode(-4)
head.next.next.next.next=temp
result=sol.detectCycle(head)
print(result.val if result!=None else "Result is None.")

File: Leetcode/142.Linked_List_CycleII/C++/142Linked_List_CycleII.cpp
#include <bits/stdc++.h> 
#include <list>
#include <map>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
struct ListNode {                           
    int val;                                
    ListNode *next;                         
    ListNode(int x) : val(x), next(NULL) {}
};
  
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
    	bool cycle=false;
    	int idx=0;
    	ListNode* ptr;
    	list<int> L1;
    	list<int>::iterator it;
    //    map<int,list<int>::iterator> mp;
    //    map<int,list<int>::iterator>::iterator it2;
    	map<int,int> mp;           
   		map<int,int>::iterator it2;   

        if(head==NULL){
            return NULL;
        }	
		else if(head->next==NULL){
			return NULL;
		}          
        if(head!=NULL){
            ptr=head;
        	L1.push_back(ptr->val);
        	it=L1.begin();
        //	mp[idx]=it;
        	mp[*it]=idx;
        	it2=mp.begin();
		}

        while(ptr->next!=NULL){
        	ptr=ptr->next;
        	L1.push_back(ptr->val);
        	cout << "ptr->val: " << ptr->val << endl;
        	idx++;
        	it++;
        	it2=mp.find(*it);
        	if(it2!=mp.end()){
        		cycle=true;
        		break;
			}       	
        	mp[*it]=idx;
		}	
		ptr=head;
		if(cycle){
			for(int i=0;i<(*it2).second;i++){
				ptr=ptr->next;
			}		
			cout << "ans: " <<"pos " << (*it2).second << " , "<<"val " << ptr->val << endl;
			return ptr;			
		}
		else{
			return NULL;
		}
    }
};

int main(){
	ListNode * headA,* headB, *ptr;
	Solution sol;
	headA=new ListNode(1);
	ptr=new ListNode(2);
	
	headA->next=ptr;
	ptr->next=headA;
//	headA=NULL;
	headB=sol.detectCycle(headA);
	if(headB==NULL){
		cout << "no cycle" << endl;
	}
	return 0;
}
